### PR TITLE
Added responseHeaders to beforeRedirect

### PR DESCRIPTION
**Pull request by pmoisel**
We would like to have access to the initial request made to read headers before redirecting to the Location-URL

### CODE REVIEW

**Comment by RubenVerborgh** [comment]
Not sure that we want to expose any internals. Maybe just a `{ headers }` object?

**Comment by pmoisel** [comment]
This would be sufficient for our needs.
I thought it would be better to expose more, so perhaps other users could use it.

I'll update the PR to just expose the header

**Comment by RubenVerborgh** [comment]
> I thought it would be better to expose more

Yes, but that might mess with the internal bookkeeping of the library. Let's not give users too much access 🙂

**Comment by oschufa** [comment]
Modifications tested to be backward compatible.

**Comment by pmoisel** [comment]
> Could you please do an option `{ headers }` as per my suggestion? Not just directly passing `headers`; the former allows for more extensibility.
> 
> Also, we will need tests.

I hope the last commit was like you asked.
Added a test using the responseHeaders.

**Comment by RubenVerborgh** [comment]
Hmm, I'll have to put this one on hold for a bit.

What we need is:
- `beforeRedirect(requestOptions, { headers })`
- a test (probably alteration to existing test) checking whether `beforeRedirect` sends `{ headers }` as a second argument
- correct documentation for that
- let's not have cookies in README or tests, that will only confuse people

File: README.md [hunk 1]
```diff
@@ -63,12 +63,24 @@ const { http, https } = require('follow-redirects');
...
+options.beforeRedirect = (options) => {
   // Use this function to adjust the options upon redirecting,
   // or to cancel the request by throwing an error
   if (options.hostname === "example.com") {
     options.auth = "user:password";
   }
+  var responseCookies = options.responseHeaders["set-cookie"];
+  if(responseCookies) {
+    var cookies = [];
+    if (optionz.headers.cookie) {
```
Comment (RubenVerborgh) [inline README.md:75]: Let's not use `optionz`

File: index.js [hunk 2]
```diff
@@ -353,6 +353,7 @@ RedirectableRequest.prototype._processResponse = function (response) {
     // Evaluate the beforeRedirect callback
     if (typeof this._options.beforeRedirect === "function") {
       try {
+        this._options.responseHeaders = this._currentRequest.res.headers;
```
Comment (RubenVerborgh) [inline index.js:356]: No need to have this in the `try`; also, should not modify the `_options` object

File: test/util.js [hunk 3]
```diff
@@ -7,6 +7,14 @@ function redirectsTo() {
   };
 }
 
+function redirectsToAndSetsCookie() {
```
Comment (RubenVerborgh) [inline test/util.js:10]: Let's keep the test simple; the cookie functionality is provided by the _test_, not the actual code, so does not need to be tested.

File: README.md (hunk 1 above)
Comment (pmoisel) [reply README.md:75]: argh. c&p-error. will be fixed

File: test/util.js (hunk 3 above)
Comment (pmoisel) [reply test/util.js:10]: ok, so test will be quite simple then...

**Comment by oschufa** [comment]
>     * let's not have cookies in README or tests, that will only confuse people

I disagree. A session cookie is the primary use case that drives the need for having access to the reponse headers of the intermediate request in the first place.

File: index.js (hunk 2 above)
Comment (pmoisel) [reply index.js:356]: I do hope I did understand you correct. See push ;)

**Comment by RubenVerborgh** [comment]
> A session cookie is the primary use case that drives the need for having access to the reponse headers of the intermediate request in the first place.

For your case, it seems to be. But the functionality as such is not specific to cookies.

**Comment by pmoisel** [comment]
Any news on this? Is the implementation now like you want it to be? Or do I have to change something?
I really hope I got your points right...
Cheers

**Comment by RubenVerborgh** [comment]
Merged as e32b25496d54d80340c6d1cb97b535e6a8a9a006, thanks for the help!

**Pull request closed without merging**